### ISSUE TITLE
Longcache should purge site cache when custom CSS is updated

### ISSUE THREAD

**Issue by jacklenox** [humanmade/longcache#1]
Our client has reported that when updating a site's Custom CSS. They incorrectly interpret this as updating a page. However, the default behaviour of Longcache probably should include a site purge when Custom CSS has been updated.

It has been many a year since I've touched anything related to Custom CSS, but it looks as though [this hook](https://developer.wordpress.org/reference/hooks/update_custom_css_data/) is the best way to achieve this.

**Comment by jacklenox** [comment]
I can probably sort a fix here this week. @joehoyle Would you mind giving me some admin rights to this repo so I can assign myself?

**Comment by joehoyle** [comment]
@jacklenox ok should be sorted!

**Comment by mattheu** [comment]
I saw this ticket and it got me thinking about this requirement, specifically how to purge the cache sitewide. Whilst his is a bit of a minor used case, I can think of many other situations you may wish to do this e.g. updating a menu. 

We can purge the CDN cache directly using a wildcard. e.g.:

```
Altis\Cloud\purge_cdn_paths( [  '/*' ] );
```

And it actually works OK if you use the longcache invalidate URL

```
Longcache\queue_invalidate_urls( [ '/*' ] );
```

First thought on this. We can't clear batcache in this way. Longcache clears that before purging the CDN cache but it needs the full URL, wildcards aren't going to work. The problem is that without doing that... it's possible that the next request will hit a page cached by batcache before the purge... and stale content will just get cached again. Workaround? Can we defer full site purges by 5 minutes. (the default max age used by batcache which as far as I can tell, remains at 300 seconds despite the max age header being set much higher, but can be modified elsewhere). Doing this would ensure the next request will hit an uncached page. Is that simple enough or am I missing something?

Second thought... what are the implications of this.  On a high traffic site, clearing everything at once could put high load on the server so probably want to avoid doing that. Maybe not something to do as standard in the plugin? But this is one of those WordPress features nobody would think about when implementing this, so perhaps there could be a setting in the Longcache plugin that can enable full site cache purges on certain actions such as this?

**Comment by joehoyle** [comment]
Yeah so really Longcache was created as a band-aid for a very specific scenario: We had a large site with an upcoming event, and there was significant possibility of the site going down, so we needed pages to be cached on the CDN in that event. Not be able to make and do site-wide flushes were a specific tradeoffs to the approach.

In recent months I've seen a couple cases of Longcache being used to assist with TTFB, however that's not really why I made it. If your TTFB is slow, the time is best spend in optimizing page generation time, not overlaying another (quite) complex layer of caching of Longcache. 

So, perhaps that's where you're coming from @mattheu, as Longcache is _not_ very productive to the day to day operations of a site. The tradeoff of the approach is: increase resilience by caching pages for a long time, in case the CMS goes offline, but give up many advantages of dynamically generated like site-wide instant updates etc.

**Comment by mattheu** [comment]
Yeah indeed that is where I'm coming from. The speed at which Cloudfront can serve a page is just so much faster than even a well optimized page, increasing the CDN cache hit rate is still appealing!

Comparing to other fairly low traffic sites, I've recently been working on a site hosted by Siteground, and they [provide a caching plugin](https://github.com/wp-plugins/sg-cachepress) that works similarly to this in that it sets a max age (24 hours by default in their case) and their servers have nginx caching that respects this. They then have some rudimentary cache clearance logic, and a purge button in the admin. What it means is that the average WordPress site is super fast without any effort at all. Having said that... we also get the odd support request related to stale content! 

I am interested in a middle ground though. 5 minute cache on a low traffic site leads to a very low cache hit rate at the CDN.  cahe Would a 1 hour cache significantly increase cache hit rates over a 5 minute one? Jack pushed one client site to 24 hours and saw the page speed reported TTFB dropped from 2s to 0.6.

**Comment by rmccue** [comment]
> Second thought... what are the implications of this. On a high traffic site, clearing everything at once could put high load on the server so probably want to avoid doing that. Maybe not something to do as standard in the plugin? But this is one of those WordPress features nobody would think about when implementing this, so perhaps there could be a setting in the Longcache plugin that can enable full site cache purges on certain actions such as this?

Each cache invalidation call has a cost associated with it, so it's not designed to be used frequently; each "path" you invalidate costs $0.005 (half a cent), where a wildcard counts as one. So, if you have a site which publishes 10 times a day, and invalidates 5 patterns for each, that's $2.50 a day - not massive, but not zero. With busier sites or more patterns, it'll add up quickly.

Cache invalidations are designed to be used in rare circumstances like an image being replaced or deleted, or as part of legal processes where content must be removed immediately, not as part of regular site operation.

(We may restrict access to the API if it's overused.)

That said, we're working on stale-while-revalidate support at the moment as AWS just added support to CloudFront, and that should have an appreciable impact on average site performance without needing super long cache lifetimes with manual validation.

**Comment by roborourke** [comment]
> each "path" you invalidate costs $0.005 (half a cent), where a wildcard counts as one. So, if you have a site which publishes 10 times a day, and invalidates 5 patterns for each, that's $2.50 a day - not massive, but not zero. With busier sites or more patterns, it'll add up quickly.

The maths is off here, it's $0.25, or 25 cents a day in that scenario. But not hard to imagine a large number of small edits could be happening on a daily basis.

**Comment by rmccue** [comment]
Sorry yeah, one order of magnitude off on that one, but my point was more that it's not within the sort of thing we should think of as free.
